### PR TITLE
fix Android CI for 32bit and 64bit builds (#10360)

### PR DESCRIPTION
**Pull request by fredowski**
* install and use NDK r21e
  * update QT install action from v2 to v3

Starting with NDK Version r24 llvm will not accept one-dash
long options anymore. This is bug QTBUG-104580

https://bugreports.qt.io/browse/QTBUG-104580

The r23c NDK version did not work due to issues with
stripping symbols. So I install and use r21e. QT 5.15.2 uses
gradle version 5.6.4 with the default android gradle plugin 3.6.0.
With gradle 6.9.2 and android gradle plugin 4.2.2 the 32bits
build worked. NDK r22b is not compatible with gstreamer 1.18 and
results in a failing 64bit build. Therefore r21e is required for
gstreamer support.

Closes: https://github.com/mavlink/qgroundcontrol/issues/10360

### CODE REVIEW

**Comment by fredowski** [comment]
I tested the 64bit and the 32bit APK on a Galaxy M20 running Android 10 and it did not crash when making a small flight plan.

**Comment by fredowski** [comment]
I tested the 64bit APK on the Galaxy M20 versus an Arduplane SITL simulation on a MacBook. I flew a simulated flight somewhere near Canberra Model Aircraft Club Flying Field.

**Comment by fredowski** [comment]
> just out of curiosity:
> 
> Why is the NDK now needed? Does gstreamer depend on it?

The NDK is always needed to build an Android application. The NDK contains compiler, linker e.t.c. The previous builds used the NDKs which are installed by github on the Ubuntu 20.04 build machine. The installed software on the Ubuntu 20.04 CI is described here:

https://github.com/actions/runner-images/blob/ubuntu20/20220821.1/images/linux/Ubuntu2004-Readme.md#android

So the latest NDK version installed on Ubuntu CI machine is currently 25.0.8775105 which is not compatible with 5.15.2 QT due to QTBUG-104580. gstreamer is precompiled with I think NDK version 21. The 1.18.6 release contained patches for NDK compatibility. Between QT and the NDK is the gradle build system connected via the "Android Gradle Plugin". The versions are either default (gradle: 5.6.4, Plugin: 3.6.0) or controlled via files in the android directory. Updating gradle and the plugin makes NDK 23 work, but NDK 24 and 25 are blocked by the QT bug. I tried gradle 6.9.2 and plugin 4.2.2. 

So I guess that simply the Ubuntu Build Machine was updated resulting in the build failure.

**Comment by zdanek** [comment]
\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\

**Comment by zdanek** [comment]
I've checked both builds - 32 and 64 bit APK on my Samsung Galaxy S9+. 
RTSP video does not work (only front and rear camera works). I don't know if it's connected with gstreamer version change or it was broken before.
I can check it but I need my work #10349 to go into master as I refurbish some parts of video code :)

**Review by zdanek — APPROVED** [review]
Works great. APKs are ok - deployed and testested both 32 and 64 on my phone.

**Comment by zdanek** [comment]
@fredowski I've checked again video on my Android. There's a problem w gstreamer. That means - it does not work. 
Either is in your build or it broke during last few commits as I had a version from about mid-august and it works good - gstreamer based video streams are available. 
So I say - **no go** for your PR until we check why video stream does not work. I should have some spare time today to check it.

**Comment by fredowski** [comment]
> @fredowski I've checked again video on my Android. There's a problem w gstreamer. That means - it does not work.

@zdanek Thank you for checking this on real hardware. gstreamer is build with NDK r21 and maybe it is important to build with the same NDK version. I have never tried video streaming and the Android device here is the phone from my son - meaning it is difficult to access test resources here... Do you need a streaming server to test the video or how do you do that?

**Comment by zdanek** [comment]
@fredowski regarding streaming server -  I will handle that. In fact I'm in a project  for VOD / Live TV streaming company so I have sufficient resources :)

**Comment by fredowski** [comment]
The build also works with r21e instead of r22b: 

https://github.com/fredowski/qgroundcontrol/pull/3

The apks are here: https://github.com/fredowski/qgroundcontrol/actions/runs/2962382845

**Comment by zdanek** [comment]
I've put 32 bit and no RTSP (and other gstreamer) video sources. Something's wrong. 
I've checked master and gstreamer video is available.

**Comment by fredowski** [comment]
> I've put 32 bit and no RTSP (and other gstreamer) video sources. Something's wrong.
> I've checked master and gstreamer video is available.

How did you check master? Did you try with a local build system?

**Comment by zdanek** [comment]
@fredowski look at build step output:

> Project MESSAGE: Skipping support for video streaming (GStreamer libraries not installed)

it cannot find GStreamer.
I got fix for you, to check, but I cannot make a PR to your fork (?)

https://github.com/mavlink/qgroundcontrol/commit/9082cdd49bf17aa2f8b754224be4f5565e6f6013

**Comment by fredowski** [comment]
@zdanek Thanks for the debugging this gstreamer problem! I added your changes to the pull request.

**Comment by zdanek** [comment]
> How did you check master? Did you try with a local build system?

most important build step is `make apk`. It uses gradle underneath. The gradle is depenedent somehow on JDK version.

This what is needed to run `make apk`:

```
JAVA_HOME=~/.sdkman/candidates/java/11.0.16-zulu/ make -j 8 apk`

```
where I point to JDK 11 - use your own.


Prior to this, I built manually with those steps - this might be  handy for someone:
in QGC directory:

```
mkdir build_android
cd build_android

qmake ../qgroundcontrol.pro -spec android-clang CONFIG+=debug CONFIG+=qml_debug ANDROID_ABIS="armeabi-v7a" 

make -j 8

mkdir android-build
make install INSTALL_ROOT=android-build

JAVA_HOME=~/.sdkman/candidates/java/11.0.16-zulu/ ~/Qt/5.15.2/android/bin/androiddeployqt --input ./android-QGroundControl-deployment-settings.json --output ./android-build  --deployment bundled --android-platform android-32  --verbose  --debug --stacktrace

```
this will build 32bit. To build 64bit ANDROID_ABIS must be set to `arm64-v8a` 
I have JDK 11 installed via sdkman. Use own version.
Also point to proper path of `androiddeployqt` if different.

I've found that crucial step is to set proper JDK to 11. I've tested 17 and 8 as I had them - no luck.
I have JDK 17 set by default, so pointing to 11 is needed.

**Comment by fredowski** [comment]
So by changing gstreamer from 1.18.5 to 1.18.6 without changing the VideoReceiver setup I simply disabled gstreamer... With the changes in VideoReceiver and hereby enabling gstreamer I see the same problems that I have seen before. The 32bit build works and the 64bit does not work. I switched the NDK from r22b to r21e and now both builds work. The default JAVA Version on the Ubuntu CI build machine is 11: https://github.com/actions/runner-images/blob/ubuntu20/20220821.1/images/linux/Ubuntu2004-Readme.md#java

Can you try the APKs with gstreamer enabled?

**Comment by zdanek** [comment]
Sure. I have some meetings. After it - I will test new APKs.

**Comment by fredowski** [comment]
I revert to gstreamer 1.18.5 and squashed the commits. We can then upgrade gstreamer from 1.18.5 to 1.18.6 in a separate commit.

**Comment by zdanek** [comment]
I've checked both 32 and 64 bit and works good with GStreaming.
Good job :medal_military:

**Comment by fredowski** [comment]
Thanks for checking and reviewing! Once this is merged I will open a new pull request for https://github.com/fredowski/qgroundcontrol/pull/4.

**Comment by zdanek** [comment]
@fredowski how can I reach you? Can you drop me a line via mail? My address in my profile. I'd like not to put it here because of spambots :D

**Comment by patrickelectric** [comment]
You can check from the apk version that the gstreamer is not working here:
7b9296bd-80b0-4498-b8b3-ac160ae0e17a
91754734-7099-48a4-8bb2-a6b495959c45

using:
`gst-launch-1.0 videotestsrc pattern=ball ! video/x-raw,width=384,height=226 ! videoconvert ! nvh264enc bitrate=5000 ! video/x-h264, profile=baseline ! h264parse ! queue ! rtph264pay config-interval=10 pt=96 ! udpsink host=192.168.0.37 port=5600gst-launch-1.0 videotestsrc pattern=ball ! video/x-raw,width=384,height=226 ! videoconvert ! nvh264enc bitrate=5000 ! video/x-h264, profile=baseline ! h264parse ! queue ! rtph264pay config-interval=10 pt=96 ! udpsink host=192.168.0.37 port=5600`

192.168.0.37 is the cellphone ip address on the network

**Comment by fredowski** [comment]
I disabled the gstreamer inclusion in the github CI build. My proposal is to fix the video streaming in separate activity as the other functions seem to work. I tested the build as mentioned above versus an Arduplane SITL session.

**Comment by fredowski** [comment]
In order to figure out why the Android video is not working I debugged the MacOS video streaming which is also not working on master and on the stable release. 

- As @AndreySkyFoxSidorov already figured out the [API warning fix](https://github.com/mavlink/qgroundcontrol/commit/3f479fc8410d3422dc4e27208b17eebcdb0bd617) resulted in non functional video in master:  https://github.com/mavlink/qgroundcontrol/issues/10378 . So I reverted that commit. I think that the code is in fact o.k. and I do not see warnings.
- The video decoding at least on MacOS does not work with the gstreamer decoder pipeline using "decodebin3". So I reverted that change which was done to improve performance. That closes: https://github.com/mavlink/qgroundcontrol/issues/10393. 
- I changed the gstreamer version for MacOS in CI from 1.18.1 to 1.18.6 because that version does not crash when decodebin3 is used - it just does not work. So 1.18.6 seems to be more stable than 1.18.1. 
- I picked the [default gstreamer for MacOS patch](https://github.com/mavlink/qgroundcontrol/commit/ef64933f1330de09c40d8622f8067d3c7be48ecb) and the [macdeployment fix](https://github.com/mavlink/qgroundcontrol/commit/8760e4a7b3d695b9ed551501fe2bad73b8b8c2bd) from @patrickelectric from the 4.2 stable branch also for master. 

With these changes I can do video streaming on Android and on MacOS. I tested this with 

`gst-launch-1.0 videotestsrc ! x264enc ! rtph264pay ! udpsink host=127.0.0.1 port=5600`

with the gstreamer 1.18.6 version installed on MacOS 10.15.7 (Catalina). 

This is the Android 64bit build running on Galaxy M20 with Android 10

IMG_7406

And this is QGroundControl on MacOS 10.15.7:

Bildschirmfoto 2022-09-05 um 01 31 19

**Comment by fredowski** [comment]
Instead of reverting the complete warning fix, I used the ideas from #10392 from @s-lisovenko. As the CI pipeline is now fully working, maybe some others could also test build results.

**Comment by pasdVn** [comment]
@fredowski I tried (for now) only the 32bit version, but it seems not to work as before (I worked with 4.2.3 before). After playing a bit around it seems as if there are problems with certain colour spaces/formats:

`gst-launch-1.0 -vvv videotestsrc ! "video/x-raw,width=320,height=240,format=(string)Y444" ! x264enc  ! h264parse ! rtph264pay! udpsink host=192.168.42.129 port=5600`
-> This works. Y444 also seems to be the default format if you don't specify the caps in your pipeline, as you did.

`gst-launch-1.0 -vvv videotestsrc ! "video/x-raw,width=320,height=240,format=(string)I420" ! x264enc ! h264parse ! rtph264pay! udpsink host=192.168.42.129 port=5600`
-> This does not work. For me it shows black screen and strange flickering overlays of other UI elments (the RTL icon or the compass rose UI, see attached video). I420 seems to be the default format when you specify the caps, but not the format (e.g. as @patrickelectric did in his pipeline).

https://user-images.githubusercontent.com/55524/188504148-d88f16c2-cd8b-45e9-8e3d-b4b8a61eb93a.mp4

If you start the first pipeline, terminate it and then start the second pipeline (before the "aiting for video" appears, still in playing state!?) the second pipeline is displayed correctly as well

With the 4.2.3 version both versions are working as expected.

**Comment by fredowski** [comment]
@pasdVn : Thank you for checking! I can reproduce the same behaviour on a Samsung Galaxy M20 phone. On MacOS all streams work. I also figured out that the recording function on Android does not work. In addition I tested an RTSP stream with `rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_115k.mp4` which works on MacOS but results in this same flickering on Android. 
Bildschirmfoto 2022-09-05 um 22 18 23

In PR #9838 the problem with the recording was already addressed. I reverted the commit 20928d20b69f only for the GstVideoReceiver.cc and GstVideoReceiver.h.

**Comment by fredowski** [comment]
The flickering on Android is gone with the last patch. Also the recording button reacts but i I have not checked the recorded videos.

**Comment by BestPrestige** [comment]
I installed the Windows installer on Windows 11 22000.856. The system uses a NVIDIA Geforce GTX 1660 SUPER graphics card. I set the stream source in QGC to "UDP h.264 Video Stream" with default video driver and tried:

`gst-launch-1.0 -vvv videotestsrc ! "video/x-raw,width=320,height=240,format=(string)Y444" ! x264enc ! h264parse ! rtph264pay! udpsink host=192.168.42.129 port=5600` => works

`gst-launch-1.0 -vvv videotestsrc ! "video/x-raw,width=320,height=240,format=(string)I420" ! x264enc ! h264parse ! rtph264pay! udpsink host=192.168.42.129 port=5600` => works

Setting the QGC Video Source to "RTSP Video Stream" with "RTSP URL" set to  

`rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_115k.mp4` => sometimes the stream is detected and sometimes not. So when I start QGC then the video is sometimes displayed and sometimes not. I attach the debug logs for the failed and the working video display.  

[failed-bunny.txt](https://github.com/mavlink/qgroundcontrol/files/9495440/failed-bunny.txt)
[failed-bunny-detailed.txt](https://github.com/mavlink/qgroundcontrol/files/9495441/failed-bunny-detailed.txt)
[working-bunny.txt](https://github.com/mavlink/qgroundcontrol/files/9495445/working-bunny.txt)
[workirg-buny-detailed.zip](https://github.com/mavlink/qgroundcontrol/files/9495457/workirg-buny-detailed.zip)

**Comment by fredowski** [comment]
I stripped all MacOS related changes such that becomes a simple patch again.

**Comment by fredowski** [comment]
@patrickelectric: I stripped all MacOS specific code and tested the 64bit apk on a Samsung M20 running Android 10.  I set the stream source in QGC to "UDP h.264 Video Stream" and tried 

`gst-launch-1.0 -vvv videotestsrc ! "video/x-raw,width=320,height=240,format=(string)Y444" ! x264enc ! h264parse ! rtph264pay! udpsink host=192.168.1.101 port=5600` => Video is displayed, Recording does not work.

`gst-launch-1.0 -vvv videotestsrc ! "video/x-raw,width=320,height=240,format=(string)I420" ! x264enc ! h264parse ! rtph264pay! udpsink host=192.168.1.101 port=5600` => Video is displayed, Recording does not work.

Setting the QGC Video Source to "RTSP Video Stream" with "RTSP URL" set to 

`rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_115k.mp4` => Video is displayed, Recording does not work

The red recording button is does not change to the rectangular shape and no video is written. I think this is https://github.com/mavlink/qgroundcontrol/pull/9838

Could you review the commit?

**Review by patrickelectric — APPROVED** [review]
Thanks @fredowski! tested and it's working!

**Comment by fredowski** [comment]
@patrickelectric : Thanks for reviewing, testing and merging!

**Comment by fredowski** [comment]
The build fails when done on the master branch where secret android keystore password is available. I fixed this with PR #10398

**Pull request merged**
